store/tikv/2pc.go

```diff
@@ -1050,3 +1008,110 @@ func appendBatchBySize(b []batchKeys, region RegionVerID, keys [][]byte, sizeFn
 	}
 	return b
 }
+
+// newBatchExecutor create processor to handle concurrent batch works(prewrite/commit etc)
+func newBatchExecutor(rateLimit int, committer *twoPhaseCommitter,
+	action twoPhaseCommitAction, procFn procOneBatchFn, backoffer *Backoffer) *batchExecutor {
+	return &batchExecutor{rateLimit, nil, committer,
+		action, procFn, backoffer}
+}
+
+// initUtils do initialize batchExecutor related policies like rateLimit util
+func (batchExe *batchExecutor) initUtils() error {
+	// init rateLimiter by injected rate limit number
+	batchExe.rateLimiter = newRateLimit(batchExe.rateLim)
+	return nil
+}
+
+// startWork concurrently do the work for each batch considering rate limit
+func (batchExe *batchExecutor) startWorker(exitCh chan struct{}, ch chan error, batches []batchKeys) {
+	for idx, batch1 := range batches {
+		if exit := batchExe.rateLimiter.getToken(exitCh); !exit {
+			batch := batch1
+			go func() {
+				var procRes error
+				defer batchExe.rateLimiter.putToken()
+				if batchExe.action == actionCommit {
+					// Because the secondary batches of the commit actions are implemented to be
+					// committed asynchronously in background goroutines, we should not
+					// fork a child context and call cancel() while the foreground goroutine exits.
+					// Otherwise the background goroutines will be canceled execeptionally.
+					// Here we makes a new clone of the original backoffer for this goroutine
+					// exclusively to avoid the data race when using the same backoffer
+					// in concurrent goroutines.
+					singleBatchBackoffer := batchExe.backoffer.Clone()
+					procRes = batchExe.procFn(singleBatchBackoffer, batch)
+				} else {
+					singleBatchBackoffer, singleBatchCancel := batchExe.backoffer.Fork()
+					defer singleBatchCancel()
+					procRes = batchExe.procFn(singleBatchBackoffer, batch)
+				}
+				ch <- procRes
+			}()
+		} else {
+			logutil.Logger(batchExe.backoffer.ctx).Info("break startWorker",
+				zap.Stringer("action", batchExe.action), zap.Int("batch size", len(batches)),
+				zap.Int("index", idx))
+			break
+		}
+	}
+}
+
+// process will start worker routine and collect results
+func (batchExe *batchExecutor) process(batches []batchKeys) error {
+	var err error
+	if len(batches) == 0 {
+		return nil
+	}
+	if len(batches) == 1 {
```

Comment:
Move this  logic to `doActionOnBatches `? 
We don't need batch executor when the len(batches) == 1.